### PR TITLE
Move font cache rebuild out of exception handler

### PR DESCRIPTION
**Pull request by Kojoley**
The purpose of the patch is to make tracebacks from cache rebuilding clearer.

### CODE REVIEW

File: lib/matplotlib/font_manager.py [hunk 1]
```diff
@@ -1356,17 +1356,20 @@ def _rebuild():
...
-            raise
-        except Exception:
-            _rebuild()
-    else:
+        except FileNotFoundError:
+            _log.debug("No font cache found %s", _fmcache)
+        except json.JSONDecodeError:
+            _log.warning("Font cache parsing failed %s", _fmcache)
+
+        if (not hasattr(fontManager, '_version') or
```
Comment (jklymak) [inline lib/matplotlib/font_manager.py]: can't `fontManager` be `None` here?  If so, won't `fontManager._version != FontManager.__version__` give an AttributeError?

Comment (Kojoley) [reply lib/matplotlib/font_manager.py]:
```
>python -c "fontManager = None; print(not hasattr(fontManager, '_version') or fontManager._version != 123)"
True
```
However `(version mismatch)` log message is misleading in this situation

Comment (jklymak) [reply lib/matplotlib/font_manager.py]: Python is so unexpected sometimes...

I guess the log message is misleading, but you'll already have had one before with the real error, so prob fine?  Or you could just check that fontManager is not None before the rest of the conditional...

Comment (Kojoley) [reply lib/matplotlib/font_manager.py]: I have updated the PR, now we should be fine.

**Comment by tacaswell** [comment]
Thanks @Kojoley !

**Pull request merged**